### DISCUSSION TITLE
Interface doesn't support nullable boolean

### DISCUSSION THREAD

**Discussion by paulboudewijn** [directus/directus#8488] [Feature Requests]
### Describe the Bug

Nullable boolean fields can't be set to null again after they have been set to true or false through the boolean interface. 
Allthough I'm fairly new to Directus, NodeJS and VueJS, I have a fix ready to solve this.

### To Reproduce

1) add a nullable boolean field to a collection
2) add an item to the collection with the boolean set to null (this works)
3) notice that in the collection list the boolean value is set to 'false' (not supposed to happen)
4) check in the database to see the value of this boolean is in fact still 'null'
5) edit this collection item; notice the checkbox is set to 'false' not 'null' (not supposed to happen)
6) click to checkbox to set the value to 'true'. save the changes.
7) edit the item again and notice the value can't be set to 'null' again. Only 'true' or 'false' are available. (not supposed to happen)

**Comment by joselcvarela** [comment]
Hello @paulboudewijn 
I've converted this into a discussion since this is more like an improvement than an issue.

Basically, from what I understood, you want or checkboxes to have tri-state: NULL, false and true.
This currently is not possible. What are your thoughts @rijkvanzanten ?

While this is currently not possible, you can use `Raw Value` (by clicking on down arrow next to field label upon hover)

**Reply by paulboudewijn** [reply]
Hi @joselcvarela,

I would indeed like checkboxes to have tri-state. 

I think at least part of the issue should be considered a bug as booleans with nullvalue in the database are rendered by Directus as 'false' instead of 'null' (both by display and interface).

Since it is possible to create nullable boolean fields through Directus, I think checkboxes should work accordingly. 

Please take a look at my solution in this branch: 
https://github.com/paulboudewijn/directus/tree/fix-nullable-boolean

I forgot to mention I'm also fairly new to github and opensource projects. So please let me know if I'm doing things wrong.

**Reply by joselcvarela** [reply]
I was already thinking if it this is also considered a bug. It might be. But since, this is already converted into a discussion let's keep here.

Can you make a pull request with that branch and refer this discussion? 
It would be great to let others check that.

**Reply by rijkvanzanten** [reply]
> I think at least part of the issue should be considered a bug as booleans with nullvalue in the database are rendered by Directus as 'false' instead of 'null' (both by display and interface).

This is indeed a bug we should resolve. It should show indeterminate for a null state.

I don't think we'd want booleans to render as a triple-toggle though. That also goes against on how a regular checkbox input behaves:

- [ ] x

Next to the raw value @joselcvarela suggested, you can nullify any field by using "Clear value" in the same field dropdown 👍🏻

**Reply by paulboudewijn** [reply]
> Can you make a pull request with that branch and refer this discussion?

Done.

**Reply by paulboudewijn** [reply]
> This is indeed a bug we should resolve. It should show indeterminate for a null state.

Thanks. Please take a look at my pull request. 

> I don't think we'd want booleans to render as a triple-toggle though. That also goes against on how a regular checkbox input behaves

Absolutely true with respect to regular checkbox behavior. But Directus does already render tri-state checkboxes for nullable booleans when inserting a _new_ item. When editing this item the boolean field becomes two-state. I'm happy to build a custom interface, but I think more users will run into this problem as nullable booleans are supported in collections. I think the way it works in my pull request is just as (non) intuitive as when inserting a new value.

**Reply by Pachat** [reply]
>  more users will run into this problem

Exactly what happent to me today. Half an hour lost to find that it is a Directus bug thanks to this post.
